packages/graph-editor/src/hooks/useCopyPaste.ts

```diff
@@ -0,0 +1,195 @@
+
+        // get the center of the visible area in flow coordinates
+        const topLeft = reactFlowInstance.screenToFlowPosition({ x: 0, y: 0 });
+        const bottomRight = reactFlowInstance.screenToFlowPosition({
+          x: window.innerWidth,
+          y: window.innerHeight,
+        });
+
+        const viewportCenter = {
+          x: (topLeft.x + bottomRight.x) / 2,
+          y: (topLeft.y + bottomRight.y) / 2,
+        };
+
+        // calculate bounding box of source nodes (these are already in flow coordinates)
+        const pasteBounds = sourceNodes.reduce(
+          (acc, node) => ({
+            minX: Math.min(acc.minX, node.annotations?.['ui.position.x'] || 0),
+            minY: Math.min(acc.minY, node.annotations?.['ui.position.y'] || 0),
+            maxX: Math.max(
+              acc.maxX,
+              (node.annotations?.['ui.position.x'] || 0) +
+                (node.annotations?.['ui.width'] || 200),
+            ),
+            maxY: Math.max(
+              acc.maxY,
+              (node.annotations?.['ui.position.y'] || 0) +
+                (node.annotations?.['ui.height'] || 100),
+            ),
+          }),
+          {
+            minX: Infinity,
+            minY: Infinity,
+            maxX: -Infinity,
+            maxY: -Infinity,
+          },
+        );
+
+        // Calculate offset to center the pasted nodes
+        const offset = {
+          x: viewportCenter.x - (pasteBounds.minX + pasteBounds.maxX) / 2,
+          y: viewportCenter.y - (pasteBounds.minY + pasteBounds.maxY) / 2,
+        };
+
+        // create new nodes with updated positions and IDs
+        const idMapping = new Map();
+        const nodes = sourceNodes.map((node) => {
+          const newId = uuidv4();
+          idMapping.set(node.id, newId);
+          return {
+            ...node,
+            id: newId,
+            annotations: {
+              ...node.annotations,
+              'ui.position.x':
+                (node.annotations?.['ui.position.x'] || 0) + offset.x,
+              'ui.position.y':
+                (node.annotations?.['ui.position.y'] || 0) + offset.y,
+            },
+          };
+        });
+
+        // create new edges with updated references
+        const edges = sourceEdges
+          .filter(
+            (edge) => idMapping.has(edge.source) && idMapping.has(edge.target),
+          )
+          .map((edge) => ({
+            ...edge,
+            id: uuidv4(),
+            source: idMapping.get(edge.source),
+            target: idMapping.get(edge.target),
+          }));
+
+        // store viewport state
+        const viewportState = reactFlowInstance.getViewport();
+
+        // merge and load new graph with selected nodes
+        const currentGraph = graphRef.getGraph().serialize();
+        const mergedGraph = {
+          ...currentGraph,
+          nodes: [...currentGraph.nodes, ...nodes].map((node) => ({
+            ...node,
+            selected: idMapping.has(node.id),
```

Comment:
We don't actually store the selected property in the serialization, this would need to be set in reactflow to handle this